File: contains-vectors-run.py
#!/usr/bin/env python3
"""
CLC-D containment vectors runner - reads containment-vectors.json
(draft-wei-clc-ext-00 §7) and runs them against the Python contains().

Asserts BOTH contains and the resolved reason code (canonical, code before
the first ':').
"""
import json
import os
import sys

from clc_semantics import contains


def canonical_reason(s):
    if not s:
        return ""
    return s.split(":", 1)[0] if ":" in s else s


def main():
    path = os.environ.get("CLC_D_VECTORS") or os.path.join(
        os.path.dirname(os.path.abspath(__file__)),
        "..", "capability", "data", "_vectors", "clc-d", "containment-vectors.json")
    with open(path) as f:
        vectors = json.load(f)

    show_all = "--all" in sys.argv
    pass_count = 0
    fail_count = 0
    for v in vectors:
        r = contains(v["parent"], v["child"])
        got_contains = r["contains"]
        got_reason = canonical_reason(r.get("reason", ""))
        want_reason = v["expect"].get("reason") or ""
        ok = got_contains == v["expect"]["contains"] and (
            got_contains or not want_reason or got_reason == want_reason)
        if ok:
            pass_count += 1
            if show_all:
                print("%-14s contains=%-5s reason=%-24s derivation=%s" % (
                    v["id"], got_contains, got_reason, v["derivation"]))
        else:
            fail_count += 1
            print("%-14s FAIL  want.contains=%-5s want.reason=%s got.contains=%-5s got.reason=%s derivation=%s" % (
                v["id"], v["expect"]["contains"], want_reason, got_contains, got_reason, v["derivation"]))

    print("Total: %d | Pass: %d | Fail: %d" % (len(vectors), pass_count, fail_count))
    return 1 if fail_count else 0


if __name__ == "__main__":
    sys.exit(main())

File: ts/contains-vectors-run.ts
// CLC-D containment vectors runner (TypeScript) — reads
// containment-vectors.json (draft-wei-clc-ext-00 §7) and runs it against the
// TS contains().  Asserts BOTH contains and the resolved reason code
// (canonical, code before the first ':').
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { contains } from './clc_semantics.ts';

const here = dirname(fileURLToPath(import.meta.url));

function canonicalReason(s: string): string {
    if (!s) return '';
    const i = s.indexOf(':');
    return i >= 0 ? s.slice(0, i) : s;
}

interface Vector {
    id: string;
    parent: { id: string; params?: Record<string, unknown> | null; constraints?: string[] };
    child: { id: string; params?: Record<string, unknown> | null; constraints?: string[] };
    expect: { contains: boolean; reason?: string };
    derivation: string;
}

const path = process.env.CLC_D_VECTORS
    || join(here, '..', '..', 'capability', 'data', '_vectors', 'clc-d', 'containment-vectors.json');
const vectors = JSON.parse(readFileSync(path, 'utf8')) as Vector[];

const showAll = process.argv.includes('--all');
let passCount = 0;
let failCount = 0;
for (const v of vectors) {
    const r = contains(v.parent, v.child);
    // No translation: contains() returns the §13.3 shape {contains, reason}
    // (rev CLC-1.15), so the runner asserts on the relation's own fields.
    const gotContains = r.contains;
    const gotReason = canonicalReason(r.reason ?? '');
    const wantReason = v.expect.reason ?? '';
    const ok = gotContains === v.expect.contains
        && (gotContains || !wantReason || gotReason === wantReason);
    if (ok) {
        passCount++;
        if (showAll) {
            console.log(`${v.id.padEnd(14)} contains=${String(gotContains).padEnd(5)} reason=${gotReason.padEnd(24)} derivation=${v.derivation}`);
        }
    } else {
        failCount++;
        console.log(`${v.id.padEnd(14)} FAIL  want.contains=${String(v.expect.contains).padEnd(5)} want.reason=${wantReason} got.contains=${String(gotContains).padEnd(5)} got.reason=${gotReason} derivation=${v.derivation}`);
    }
}
console.log(`Total: ${vectors.length} | Pass: ${passCount} | Fail: ${failCount}`);
process.exit(failCount ? 1 : 0);